### PR TITLE
Regular sequence: closing experimental phase of regular sequence module

### PR DESCRIPTION
**Pull request by dkrenn**
This PR closes the experimental-phase of the regular sequences module. This module has been experimental für a very long time now. Lately #35894 has been completed so this is now a fully-functional ring/algebra. Except for the consequences of  #35894 (renaming; see below), the interface is stable for very long time.

This PR does:
- Rename to `RegularSequenceRing` (as this structure now is a ring due to #35894).
- Remove the experimental warning.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### :hourglass: Dependencies

- #35894

### CODE REVIEW

**Comment by dkrenn** [comment]
CC @cheuberg

File: src/sage/combinat/all.py [hunk 1]
```diff
@@ -277,7 +277,7 @@
 lazy_import('sage.combinat.binary_recurrence_sequences',
             'BinaryRecurrenceSequence')
 lazy_import('sage.combinat.recognizable_series', 'RecognizableSeriesSpace')
-lazy_import('sage.combinat.k_regular_sequence', 'kRegularSequenceSpace')
+lazy_import('sage.combinat.k_regular_sequence', 'RegularSequenceRing')
```
Comment (cheuberg) [inline src/sage/combinat/all.py]: I propose to rename the file `k_regular_sequence.py` to `regular_sequence.py` (after merging #36001) for consistency: now that everything is renamed in such a way to remove the variable `k`, that would seem to be a logical move to me.

File: src/sage/combinat/k_regular_sequence.py [hunk 2]
```diff
@@ -6,21 +6,6 @@
 found, for example, on the :wikipedia:`k-regular_sequence` or in
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:6]: The `k` in  lines 3 and 5 could be removed.

File: src/sage/combinat/k_regular_sequence.py [hunk 3]
```diff
@@ -188,14 +173,14 @@ class DegeneratedSequenceError(RuntimeError):
     pass
 
 
-class kRegularSequence(RecognizableSeries):
+class RegularSequence(RecognizableSeries):
     def __init__(self, parent, mu, left=None, right=None):
         r"""
         A `k`-regular sequence.
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:179]:
```suggestion
        A regular sequence.
```

File: src/sage/combinat/k_regular_sequence.py [hunk 4]
```diff
@@ -245,7 +230,7 @@ def __init__(self, parent, mu, left=None, right=None):
         .. SEEALSO::
 
             :doc:`k-regular sequence <k_regular_sequence>`,
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py]:
```suggestion
            :doc:`regular sequence <k_regular_sequence>`,
```

File: src/sage/combinat/k_regular_sequence.py [hunk 5]
```diff
@@ -973,15 +958,15 @@ def backward_differences(self, **kwds):
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:958]: In the one-line description, there is a `k` which could be removed.

File: src/sage/combinat/k_regular_sequence.py [hunk 6]
```diff
@@ -1037,6 +1022,115 @@ def forward_differences(self, **kwds):
         """
         return self.subsequence(1, {1: 1, 0: -1}, **kwds)
 
+    @minimize_result
+    def _mul_(self, other):
+        r"""
+        Return the product of this `k`-regular sequence with ``other``,
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:1028]:
```suggestion
        Return the product of this regular sequence with ``other``,
```

File: src/sage/combinat/k_regular_sequence.py [hunk 7]
```diff
@@ -1014,11 +999,11 @@ def forward_differences(self, **kwds):
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:999]: In the one-line description, a `k` could be removed.

File: src/sage/combinat/k_regular_sequence.py [hunk 8]
```diff
@@ -357,7 +342,7 @@ def is_degenerated(self):
```
Comment (cheuberg) [inline src/sage/combinat/k_regular_sequence.py:342]: In the one-sentence description, two `k` could be removed.

**Review by cheuberg — CHANGES_REQUESTED** [review]
The changes so far are fine with me, but I'd go farther with removing the name `k` of the base. I started making suggestions but then stopped because a search-and-replace for "`k`-regular" seems to be more efficient.

File: src/sage/combinat/all.py (hunk 1 above)
Comment (dkrenn) [reply src/sage/combinat/all.py:280]: Agreed.

Comment (dkrenn) [reply src/sage/combinat/all.py]: Yes, indeed, I agree that this seems a logical move. I prepared a commit and will push it soon. (BTW it does not seem to be a problem due to #36001 when merging; `git` appearently takes care of this.)

**Comment by dkrenn** [comment]
> The changes so far are fine with me, but I'd go farther with removing the name `k` of the base.

At the moment, I am not fully convinced that this is the way to go. After all, the used term is "`k`-regular sequence"; see titles of the corresponding articles and also the Wikipedia article https://en.wikipedia.org/wiki/K-regular_sequence consistently uses this term including the base. (FWIW The wikipedia article https://en.wikipedia.org/wiki/Regular_sequence refers to a different concept, but at least in the first line k-regular sequences are referenced; as our file is in `sage.combinat`, I do not see a problem with the renaming of the file as discussed above.) Moreover, I do not see a problem that the class `RegularSequence` represents a `k`-regular sequence because from a coding perspective it seems clear that the parameter `k` has to be fixed somewhere (but not in the name of the class).

I, therefore, have some preference, for keeping this as it is.

In case we agree to change, I have a strong preference for keeping the title (line 3) of the module as it is (i.e. including the base) and think we should give both names `k`-regular and regular (and say something that we will simply skip the base) in the paragraph starting with line 5.

File: src/sage/combinat/k_regular_sequence.py (hunk 2 above)
Comment (dkrenn) [reply src/sage/combinat/k_regular_sequence.py:6]: Not convinced; see below (https://github.com/sagemath/sage/pull/36011#issuecomment-1659826240)

**Review by cheuberg — APPROVED** [review]
Changes are fine with me. The question on how to call the sequences in the documentation is perhaps not very urgent (and independent of removing the experimental notice).

Two comments:

- while the Wikipedia entry for regular sequences is called "k-regular sequence", the entry there for the closely related automatic sequence is called "automatic sequence" with an explanation there that it is sometimes called a "k-automatic sequence".
- concerning the documentation here: for some methods, mentioning "k" makes more sense to me than for others. E.g. for the constructor of a regular sequence, it makes more sense to me to say that it is "k-recursive" because the description of the family mu refers to the same k. However, for the oneline description of backwards difference, the "k" seems to be quite redundant to me: it does not occur anywhere else in the description of the method.

But, as I said, we can also merge it essentially as it is, perhaps just adding a short sentence at the very beginning.

**Comment by dkrenn** [comment]
> Changes are fine with me. The question on how to call the sequences in the documentation is perhaps not very urgent (and independent of removing the experimental notice).

This is now #36019.

**Pull request merged**